### PR TITLE
005 view comps by band

### PR DESCRIPTION
**Pull request by stuferg**

### CODE REVIEW

File: src/main/java/com/kainos/ea/model/CompetencyRequest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+public class CompetencyRequest {
+
+
+    private String competencyName;
+    private String description;
+    private String bandName;
+
+
+    public CompetencyRequest(String competencyName, String description, String bandName) {
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/model/CompetencyRequest.java:11]: Should this have the @JSONProperty annotation?

```suggestion
    public CompetencyRequest(String competencyName, String description, String bandName) {
```

File: src/main/java/com/kainos/ea/service/CompetencyService.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+
+}
+
+
+
+
+
+
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/service/CompetencyService.java:45]: Do we need this blank space?

File: SQLScripts/US005-test-data.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    VALUES ("Associate");
+    
+    INSERT INTO Bands(band_name)
+    VALUES ("Senior Associate");
+    
+    INSERT INTO Bands(band_name)
+    VALUES ("Consultant");
+    
+    INSERT INTO Bands(band_name)
+    VALUES ("Manager");
```
[the remark this answers is not in the document]
Comment (stuferg) [reply SQLScripts/US005-test-data.sql:14]: Done

File: SQLScripts/US005-test-data.sql [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    VALUES ("Setting Direction, Development and Accountability", "Effective Leadership require individuals to contribute");
+    
+    INSERT INTO Competencies(name, description)
+    VALUES ("Supporting and Delivering the Strategy", "Effective Leadership require support, contribution, and delivery");
+    
+    INSERT INTO Competencies(name, description)
+    VALUES ("Commerciablity and Risk", "Effective Leadership require individuals to demonstrate effectivness");
+    
+    INSERT INTO Competencies(name, description)
+    VALUES ("Communicating and Influence", "Effective Leadership require individuals to be exceptional");
```
[the remark this answers is not in the document]
Comment (stuferg) [reply SQLScripts/US005-test-data.sql:32]: I just wanted this for test data, but made the changes

File: SQLScripts/US005.sql [hunk 5]
```diff
@@ -0,0 +1,24 @@
+	CREATE TABLE IF NOT EXISTS Competencies (
+competency_id TINYINT AUTO_INCREMENT,
+name varchar(50),
+description varchar (500),
+PRIMARY KEY (competency_id)
+);
```
[the remark this answers is not in the document]
Comment (stuferg) [reply SQLScripts/US005.sql:6]: I chatted to the team and checked the tickets and we think this is the first one which uses competencies. Which ticket were you thinking of?

File: SQLScripts/US005.sql [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+band_competency_id TINYINT AUTO_INCREMENT NOT NULL,
+band_id TINYINT NOT NULL,
+competency_id TINYINT NOT NULL,
+PRIMARY KEY (band_competency_id)
+);
+
+ALTER TABLE Band_Competencies
+ADD CONSTRAINT fk_Band_Competencies_BandId
+FOREIGN KEY(band_id)
+REFERENCES Bands(band_id);
```
[the remark this answers is not in the document]
Comment (stuferg) [reply SQLScripts/US005.sql:18]: Done

File: SQLScripts/US005.sql [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+
+ALTER TABLE Band_Competencies
+ADD CONSTRAINT fk_Band_Competencies_CompetencyId
+FOREIGN KEY(competency_id)
+REFERENCES Competencies(competency_id);
```
[the remark this answers is not in the document]
Comment (stuferg) [reply SQLScripts/US005.sql:24]: Done

File: src/main/java/com/kainos/ea/model/CompetencyRequest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package com.kainos.ea.model;
+
+public class CompetencyRequest {
```
[the remark this answers is not in the document]
Comment (stuferg) [reply src/main/java/com/kainos/ea/model/CompetencyRequest.java:3]: Refactored

File: src/test/java/com/kainos/ea/controller/CompetencyControllerTests.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        CompetencyService compService = Mockito.mock(CompetencyService.class);
+
+        CompetencyController compController = new CompetencyController(compService);
+
+        Connection conn;
+
+        static final DropwizardAppExtension<WebServiceConfiguration> APP = new DropwizardAppExtension<>(
+                WebServiceApplication.class, null,
+                new ResourceConfigurationSourceProvider()
+        );
```
[the remark this answers is not in the document]
Comment (stuferg) [reply src/test/java/com/kainos/ea/controller/CompetencyControllerTests.java:30]: I've taken it out

File: src/test/java/com/kainos/ea/integration/CompetencyIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            new ResourceConfigurationSourceProvider());
+
+    @Test
+    void getCompsByBand_shouldReturnAListOfComps() {
+        int bandId = 1;
+        List<CompetencyRequest> response = APP.client().target("http://localhost:8080/api/competencies/" + bandId)
+                .request()
+                .get(List.class);
+
+        Assertions.assertTrue(response.size()>0);
```
[the remark this answers is not in the document]
Comment (stuferg) [reply src/test/java/com/kainos/ea/integration/CompetencyIntegrationTest.java:29]: As in more tests? Can do

File: .gitignore [hunk 11]
```diff
@@ -26,4 +26,6 @@ target/*
 
 .idea/workspace.xml
```
Comment (ccurran01) [inline .gitignore:27]: Think this should be .idea/. Seems that some files from the .idea folder have slipped past this check

File: .idea/misc.xml
Comment (ccurran01) [inline .idea/misc.xml:1]: Could this be removed?

File: SQLScripts/US005.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (ccurran01) [reply SQLScripts/US005.sql:6]: I might be loosing the plot, no worries

File: .idea/misc.xml
Comment (stuferg) [reply .idea/misc.xml:1]: Deleting this gives me some strange errors, not sure if you'd like to have a look?

File: .gitignore [hunk 12]
```diff
@@ -24,6 +24,8 @@ hs_err_pid*
 
 target/*
 
-.idea/workspace.xml
+.idea/.
```
Comment (ccurran01) [inline .gitignore:27]: `git rm ./idea -r` should clear up all these files

```suggestion
.idea/*
```

File: JavaWebService.iml
Comment (ccurran01) [inline JavaWebService.iml:1]: This should also be added to the git ignore. a `git rm JavaWebService.iml` should stop it from being tracked in the future as well

File: src/main/java/com/kainos/ea/model/CompetencyResponse.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        return bandName;
+    }
+
+    public void setBandName(String bandName) {
+        this.bandName = bandName;
+    }
+}
+
+
+
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/model/CompetencyResponse.java:49]: You have a pet cat I don't know about @stuferg? 😆 

https://media.giphy.com/media/26n6xF5M2Ht4eKdO0/giphy.gif

Don't think we need this white space

File: src/test/java/com/kainos/ea/service/CompetencyServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(MockitoExtension.class)
+public class CompetencyServiceTest {
```
Comment (ccurran01) [inline src/test/java/com/kainos/ea/service/CompetencyServiceTest.java:20]: Can we add another scenario here for when a database connection exception is thrown

File: src/test/java/com/kainos/ea/service/CompetencyServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    @Test
+    void getCompsWithBand_shouldThrowFailedToGetCompsException_whenDaoThrowsSQLException() throws SQLException, DatabaseConnectionException {
+        Mockito.when(dbConnector.getConnection()).thenReturn(connection);
+        Mockito.when(compDao.getAllCompsAndBand(bandId, connection)).thenThrow(SQLException.class);
+
+        assertThrows(FailedToGetCompsException.class,
+                () -> compService.getAllCompsWithBand(bandId));
+    }
+
+
```
Comment (ccurran01) [inline src/test/java/com/kainos/ea/service/CompetencyServiceTest.java:47]: Some extra blank space here also

**Review by ccurran01 — COMMENTED** [review]
Good work @stuferg, nearly there

**Review by ccurran01 — APPROVED** [review]

**Pull request merged**